README.md
# rolling_nav_bar

A bottom nav bar with layout inspired by [this design](https://dribbble.com/shots/5713130-Greeny-app) and with heavily customizable animations, colors, and shapes.

## Getting Started

To get started, place your `RollingNavBar` in the `bottomNavigationCar` slot of a
`Scaffold`, wrapped in a widget that provides max height. For example:

```dart
Scaffold(
  bottomNavigationBar: Container(
    height: 95,
    child: RollingNavBar(
      // nav items
    ),
  )
);
```

Alternatively, you can place it directly using a `Stack`:

```dart
Scaffold(
  body: Stack(
    children: <Widget>[
      Positioned(
        bottom: 0,
        height: 95,
        width: MediaQuery.of(context).size.width,
        child: RollingNavBar(
          // nav items
        ),
      )
    ]
  )
);
```

## Customization

`RollingNavBar` is heavily customizable and works with 3, 4, or 5 navigation elements.
Each element is also fully customizable through the two primary ways to specify child
navigation elements.

The first option is to pass a list of `IconData` objects, along with optional lists
of `Color` objects.

```dart
RollingNavBar.iconData(
  iconData: <IconData>[
    Icons.home,
    Icons.people,
    Icons.settings,
  ],
  indicatorColors: <Color>[
    Colors.red,
    Colors.yellow,
    Colors.blue,
  ],
)
```

The second option is to provide a widget builder, though this comes with some loss of
helpful active state management.

```dart
RollingNavBar.builder(
  builder: (context, index, info, update) {
    // The `iconData` constructor handles all active state management,
    // but this constructor, as it deals with completed widgets, must
    // assume that you have already made all relevant considerations.
    var textStyle = index == info.nextIndex
      ? TextStyle(color: Colors.white)
      : TextStyle(color: Colors.grey);
    return Text('${index + 1}', style: style);
  },
  indicatorColors: <Color>[
    Colors.red,
    Colors.yellow,
    Colors.blue,
  ],
  numChildren: 3,
)
```

## Animation Types

`RollingNavBar` comes with four animation flavors for the active indicator's transition from
tab to tab.

The first animation type is the namesake: `AnimationType.roll`:

```dart
RollingNavBar.iconData(
  animationCurve: Curves.easeOut,  // `easeOut` (the default) is recommended here
  animationType: AnimationType.roll,
  baseAnimationSpeed: 200, // milliseconds
  iconData: <IconData>[
    ...
  ],
)
```

<img src="https://raw.githubusercontent.com/craiglabenz/flutter_rolling_nav_bar/master/doc/assets/roll.gif" alt="Roll Example" height="600" />

<br>

> Note: For the `roll` animation type, your supplied animation speed is a multiplier considered against the distance the indicator must travel. This ensures a constant speed of travel no matter where the user clicks.

---

The second animation type is a fade-and-reappear effect:

```dart
RollingNavBar.iconData(
  animationCurve: Curves.linear, // `linear` is recommended for `shrinkOutIn`
  animationType: AnimationType.shrinkOutIn,
  baseAnimationSpeed: 500, // slower animations look nicer for `shrinkOutIn`
  iconData: <IconData>[
    ...
  ],
)
```

<img src="https://raw.githubusercontent.com/craiglabenz/flutter_rolling_nav_bar/master/doc/assets/shrink-out-in.gif" alt="Shrink-out-in Example" height="600" />

<br>

> Note: For the `shinkOutIn` animation type, your supplied animation speed is constant, since the active indicator never travels the intermediate distance.

---

The third animation type is a spinning version of fade-and-reappear:

```dart
RollingNavBar.iconData(
  animationCurve: Curves.linear, // `linear` is recommended for `spinOutIn`
  animationType: AnimationType.spinOutIn,
  baseAnimationSpeed: 500, // slower animations look nicer for `spinOutIn`
  iconData: <IconData>[
    ...
  ],
)
```

<img src="https://raw.githubusercontent.com/craiglabenz/flutter_rolling_nav_bar/master/doc/assets/spin-out-in.gif" alt="Spin-out-in Example" height="600" />

<br>

> Note: Like with `shinkOutIn`, for the `spinOutIn` animation type, your supplied animation speed is constant, since the active indicator never travels the intermediate distance.

---

The final animation flavor is a non-animation:

```dart
RollingNavBar.iconData(
  // `animationCurve` and `baseAnimationSpeed` are ignored
  animationType: AnimationType.snap,
  iconData: <IconData>[
    ...
  ],
)
```

<img src="https://raw.githubusercontent.com/craiglabenz/flutter_rolling_nav_bar/master/doc/assets/snap.gif" alt="Snap Example" height="600" />

<br>

## Hooking into the animation

In the demo, the background of the larger hexagon matches the background of
the nav bar hexagon. To achieve this and similar effects, two callbacks, `onTap` and
`onAnimate`, are available. `onAnimate` can be particularly helpful for syncing
visual effects elsewhere in your app with nav bar progress.

## Tab Item Text

If using the `iconData` constructor, you are also able to pass a list of widgets
to render as text below inactive icons.

```dart
RollingNavBar.iconData(
  // A list of length one implies the same color for all icons
  iconColors: <Color>[
    Colors.grey[800],
  ],
  iconData: <IconData>[
    Icons.home,
    Icons.friends,
    Icons.settings,
  ],
  iconText: <Widget>[
    Text('Home', style: TextStyle(color: Colors.grey, fontSize: 12)),
    Text('Friends', style: TextStyle(color: Colors.grey, fontSize: 12)),
    Text('Settings', style: TextStyle(color: Colors.grey, fontSize: 12)),
  ]
)
```

<img src="https://raw.githubusercontent.com/craiglabenz/flutter_rolling_nav_bar/master/doc/assets/iconText.png" alt="Icon Text Example" height="600" />

## Icon badges

Using the [Badges](https://pub.dev/packages/badges) library, `RollingNavBar` is able to easily expose nav bar badges. The following works with either constructor.

```dart
RollingNavBar.iconData(
  badges: <Widget>[
    Text('1', style: TextStyle(Colors.white)),
    Text('1', style: TextStyle(Colors.white)),
    null,
    null,
    Text('1', style: TextStyle(Colors.white)),
  ],
  iconData: <IconData>[
    Icons.home,
    Icons.friends,
    Icons.account,
    Icons.chat,
    Icons.settings,
  ],
```

<img src="https://raw.githubusercontent.com/craiglabenz/flutter_rolling_nav_bar/master/doc/assets/badges.png" alt="Badges Example" height="600" />

## Driving Navigation Bar Changes

You can programmatically change the active navigation bar tab by passing a new `activeIndex` to the `RollingNavBar` widget. However, there are two steps to successfully keeping everything in sync.

```dart
class _MyAppState extends State<MyApp> {
  int activeIndex;

  /// Handler that responds to navigation events (likely derived
  /// from user clicks) and keeps the parent in sync.
  void _onTap(int index) {
    setState(() {
      activeIndex = index;
    });
  }

  /// Handler for when you want to programmatically change
  /// the active index. Calling `setState()` here causes
  /// Flutter to re-render the tree, which `RollingNavBar`
  /// responds to by running its normal animation.
  ///
  /// Note: This is an example function that you can define
  /// and call wherever you like. That is why it is not passed
  /// to `RollingNavBar.iconData` in the `build` method.
  void changeActiveIndex(int index) {
    setState((){
      activeIndex = index;
    });
  }

  Widget build(BuildContext context) {
    return RollingNavBar.iconData(
      activeIndex: activeIndex,
      iconData: iconData,
      onTap: _onTap,
    );
  }
}
```


CHANGELOG.md
## [0.0.5-nullsafety.3] - 2021-05-10

- Updated available dependencies to null-safe versions

## [0.0.5-nullsafety.1] - 2021-02-18

- Added null-safety declarations

## [0.0.4] - 2021-02-17

- Fixed problem where a supplied `onTap()` callback could not call `setState()`

## [0.0.3] - 2020-12-19

- Ensured onTap callback always fires, even if index is current

## [0.0.2] - 2020-08-08

- Added RTL support
- Verified functionality with Flutter 1.20

## [0.0.1] - 2020-01-07

- Removed `children` constructor in favor of `builder` pattern

## [0.0.1-alpha+8] - 2019-12-28

- Added customization for badge colors

## [0.0.1-alpha+7] - 2019-12-28

- Added support for icon badges

## [0.0.1-alpha+6] - 2019-12-19

- Fixed typos. Functionally equivalent to 0.0.1-alpha+5

## [0.0.1-alpha+5] - 2019-12-19

- Added ability to programmatically drive new nav bar index

## [0.0.1-alpha+4] - 2019-12-17

- Fixed bug with null error when not passing `iconText`
- Added general sanity tests

## [0.0.1-alpha-3] - 2019-12-17

- Added support for icon text

## [0.0.1-alpha-2] - 2019-12-17

- More README clarifications

## [0.0.1-alpha-1] - 2019-12-17

- Variable name clean-up and other documentation enhancements

## [0.0.1-alpha] - 2019-12-15

- Initial release of `RollingNavBar` with four animation types: roll, snap, fade-out-and-in, and spin-out-and-in


test/clicks_test.dart
import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';

import './stateful_test_widget.dart';

void main() {
  Widget _rollingNavBarApp([Function(int)? onTap]) {
    return MaterialApp(
      home: StatefulTestWidget(onTap),
    );
  }

  group('RollingNavBar renders correctly', () {
    testWidgets(
      'with on an ontap that doesn\'t call setState',
      (WidgetTester tester) async {
        final app = _rollingNavBarApp();
        await tester.pumpWidget(app);
        await tester.tap(find.byKey(Key('nav-bar-child-2')));
        await tester.pump();
        // Not just that the onTap callback was fired,
        // but that the new stuff renders.
        expect(find.text('Active Index: 2'), findsOneWidget);
      },
    );

    testWidgets(
      'with calls onTap when tapped index is already active',
      (WidgetTester tester) async {
        bool callbackFired = false;
        final app = _rollingNavBarApp((_) => callbackFired = true);
        await tester.pumpWidget(app);
        await tester.tap(find.byKey(Key('nav-bar-child-0')));
        await tester.pump();
        // Not just that the onTap callback was fired,
        // but that the new stuff renders.
        expect(callbackFired, true);
      },
    );

    testWidgets(
      'with on separate action that forces a new index',
      (WidgetTester tester) async {
        final app = _rollingNavBarApp();
        await tester.pumpWidget(app);
        // Taps on the FAB increment the index
        await tester.tap(find.byKey(Key('FAB')));
        await tester.pump();
        // Not just that the onTap callback was fired,
        // but that the new stuff renders.
        expect(find.text('Active Index: 1'), findsOneWidget);
      },
    );
  });
}


test/rolling_nav_bar_test.dart
import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';

import 'package:rolling_nav_bar/rolling_nav_bar.dart';

void main() {
  // Helper function to remove lots of meaningless lines of code from
  // appearing and test after test below
  Widget _materialApp(
    RollingNavBar navBar, {
    double height: 100,
    double width: 400,
  }) {
    return MaterialApp(
      home: Scaffold(
        bottomNavigationBar: Container(
          child: navBar,
          height: height,
          width: width,
        ),
      ),
    );
  }

  group('RollingNavBar builds correctly', () {
    testWidgets('with simplest iconData parameters',
        (WidgetTester tester) async {
      await tester.pumpWidget(
        _materialApp(
          RollingNavBar.iconData(
            iconData: <IconData>[Icons.home, Icons.person, Icons.settings],
          ),
        ),
      );
      expect(find.byIcon(Icons.home), findsOneWidget);
      expect(find.byIcon(Icons.person), findsOneWidget);
      expect(find.byIcon(Icons.settings), findsOneWidget);
    });

    testWidgets('with simplest builder parameters',
        (WidgetTester tester) async {
      final textWidgets = <Widget>[
        Text('Home'),
        Text('Friends'),
        Text('Settings')
      ];
      await tester.pumpWidget(
        _materialApp(
          RollingNavBar.builder(
            builder: (
              BuildContext context,
              int index,
              AnimationInfo? info,
              AnimationUpdate? update,
            ) =>
                textWidgets[index],
            numChildren: 3,
          ),
        ),
      );
      expect(find.text('Home'), findsOneWidget);
      expect(find.text('Friends'), findsOneWidget);
      expect(find.text('Settings'), findsOneWidget);
    });

    testWidgets('with builder onTap', (WidgetTester tester) async {
      int? tapped;
      await tester.pumpWidget(
        _materialApp(
          RollingNavBar.builder(
            builder: (
              BuildContext context,
              int index,
              AnimationInfo? info,
              AnimationUpdate? update,
            ) =>
                Text(index.toString(), key: Key('key-$index')),
            numChildren: 3,
            onTap: (int index) {
              tapped = index;
            },
          ),
        ),
      );
      await tester.tap(find.byKey(Key('key-1')));
      expect(tapped, 1);
    });

    testWidgets('with iconData onTap', (WidgetTester tester) async {
      int? tapped;
      await tester.pumpWidget(
        _materialApp(
          RollingNavBar.iconData(
            iconData: <IconData>[
              Icons.home,
              Icons.person,
              Icons.settings,
            ],
            onTap: (int index) {
              tapped = index;
            },
          ),
        ),
      );
      await tester.tap(find.byKey(Key('nav-bar-child-2')));
      expect(tapped, 2);
    });

    testWidgets('with iconData and text', (WidgetTester tester) async {
      await tester.pumpWidget(
        _materialApp(
          RollingNavBar.iconData(
            iconData: <IconData>[Icons.home, Icons.person, Icons.settings],
            iconText: <Widget>[Text('Home'), Text('Friends'), Text('Settings')],
          ),
        ),
      );
      expect(find.byIcon(Icons.home), findsOneWidget);
      expect(find.byIcon(Icons.person), findsOneWidget);
      expect(find.byIcon(Icons.settings), findsOneWidget);
      // Finds nothing because is starting active widget
      expect(find.text('Home'), findsNothing);
      // Inactive tabs render their iconText
      expect(find.text('Friends'), findsOneWidget);
      expect(find.text('Settings'), findsOneWidget);
    });
    testWidgets('with iconData and text and specified starting active index',
        (WidgetTester tester) async {
      await tester.pumpWidget(
        _materialApp(
          RollingNavBar.iconData(
            activeIndex: 1,
            iconData: <IconData>[Icons.home, Icons.person, Icons.settings],
            iconText: <Widget>[Text('Home'), Text('Friends'), Text('Settings')],
          ),
        ),
      );
      // Inactive tabs render their iconText
      expect(find.text('Home'), findsOneWidget);
      // Finds nothing because is starting active widget
      expect(find.text('Friends'), findsNothing);
      expect(find.text('Settings'), findsOneWidget);
    });
    testWidgets('with a custom active indicator color',
        (WidgetTester tester) async {
      await tester.pumpWidget(
        _materialApp(
          RollingNavBar.iconData(
            activeIndex: 1,
            iconData: <IconData>[Icons.home, Icons.person, Icons.settings],
            iconText: <Widget>[Text('Home'), Text('Friends'), Text('Settings')],
            indicatorColors: <Color>[Colors.orange],
          ),
        ),
      );
      // Inactive tabs render their iconText
      Finder activeIndicator = find.byType(ActiveIndicator);
      expect(activeIndicator, findsOneWidget);

      final activeIndicatorWidget =
          tester.firstWidget(activeIndicator) as ActiveIndicator;
      expect(activeIndicatorWidget.color, Colors.orange);
    });
  });
}


test/rolling_tab_bar_test.dart
// import 'package:flutter_test/flutter_test.dart';

// import 'package:rolling_tab_bar/rolling_tab_bar.dart';

void main() {
  // test('adds one to input values', () {
  //   final calculator = Calculator();
  //   expect(calculator.addOne(2), 3);
  //   expect(calculator.addOne(-7), -6);
  //   expect(calculator.addOne(0), 1);
  //   expect(() => calculator.addOne(null), throwsNoSuchMethodError);
  // });
}


example/lib/main.dart
import 'dart:ui';
import 'package:flutter/material.dart';
import 'package:flutter_polygon/flutter_polygon.dart';
import 'package:rolling_nav_bar/indexed.dart';
import 'package:rolling_nav_bar/rolling_nav_bar.dart';

void main() => runApp(MyApp());

double scaledHeight(BuildContext context, double baseSize) {
  return baseSize * (MediaQuery.of(context).size.height / 800);
}

double scaledWidth(BuildContext context, double baseSize) {
  return baseSize * (MediaQuery.of(context).size.width / 375);
}

class MyApp extends StatefulWidget {
  // This widget is the root of your application.
  @override
  _MyAppState createState() => _MyAppState();
}

class _MyAppState extends State<MyApp> {
  Color logoColor = Colors.red[600]!;
  int activeIndex = 0;

  var iconData = <IconData>[
    Icons.home,
    Icons.people,
    Icons.account_circle,
    Icons.chat,
    Icons.settings,
  ];

  var badges = <int>[0, 0, 0, 0, 0];

  var iconText = <Widget>[
    Text('Home', style: TextStyle(color: Colors.grey, fontSize: 12)),
    Text('Friends', style: TextStyle(color: Colors.grey, fontSize: 12)),
    Text('Account', style: TextStyle(color: Colors.grey, fontSize: 12)),
    Text('Chat', style: TextStyle(color: Colors.grey, fontSize: 12)),
    Text('Settings', style: TextStyle(color: Colors.grey, fontSize: 12)),
  ];

  var indicatorColors = <Color>[
    Colors.red,
    Colors.orange,
    Colors.green,
    Colors.blue,
    Colors.purple,
  ];

  List<Widget?> get badgeWidgets => indexed(badges)
      .map((Indexed indexed) => indexed.value > 0
          ? Text(indexed.value.toString(),
              style: TextStyle(
                color: indexed.index == activeIndex
                    ? indicatorColors[indexed.index]
                    : Colors.white,
              ))
          : null)
      .toList();

  void incrementIndex() {
    setState(() {
      activeIndex = activeIndex < (iconData.length - 1) ? activeIndex + 1 : 0;
    });
  }

  _onAnimate(AnimationUpdate update) {
    setState(() {
      logoColor = update.color;
    });
  }

  _onTap(int index) {
    if (activeIndex == index) {
      _incrementBadge();
    }
    activeIndex = index;
    setState(() {});
  }

  void _incrementBadge() {
    setState(() {
      badges[activeIndex] += 1;
    });
  }

  List<Widget> get builderChildren => const <Widget>[
        Text('1', style: TextStyle(color: Colors.grey)),
        Text('2', style: TextStyle(color: Colors.grey)),
        Text('3', style: TextStyle(color: Colors.grey)),
      ];

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      theme: ThemeData(
        scaffoldBackgroundColor: Colors.blue[100],
      ),
      home: Builder(
        builder: (BuildContext context) {
          double largeIconHeight = MediaQuery.of(context).size.width;
          double navBarHeight = scaledHeight(context, 85);
          double topOffset = (MediaQuery.of(context).size.height -
                  largeIconHeight -
                  MediaQuery.of(context).viewInsets.top -
                  (navBarHeight * 2)) /
              2;
          return Scaffold(
            floatingActionButton: FloatingActionButton(
              backgroundColor: logoColor,
              child: Icon(Icons.add),
              onPressed: _incrementBadge,
            ),
            appBar: AppBar(
              title: Text('Rolling Nav Bar: Tab ${activeIndex + 1}'),
            ),
            body: Stack(
              children: <Widget>[
                Positioned(
                  top: topOffset,
                  height: largeIconHeight,
                  width: largeIconHeight,
                  child: GestureDetector(
                    onTap: incrementIndex,
                    child: ClipPolygon(
                      sides: 6,
                      borderRadius: 15,
                      child: Container(
                        height: largeIconHeight,
                        width: largeIconHeight,
                        color: logoColor,
                        child: Center(
                          child: Padding(
                            padding: EdgeInsets.fromLTRB(0, 100, 30, 0),
                            child: Transform(
                              transform: Matrix4.skew(0.1, -0.50),
                              child: Text(
                                'Rolling\nNav Bar',
                                textAlign: TextAlign.center,
                                style: TextStyle(
                                  color: Colors.white,
                                  fontSize: scaledWidth(context, 63),
                                  fontFeatures: <FontFeature>[
                                    FontFeature.enable('smcp')
                                  ],
                                  shadows: <Shadow>[
                                    Shadow(
                                      offset: Offset(5, 5),
                                      blurRadius: 3.0,
                                      color: Color.fromARGB(255, 0, 0, 0),
                                    ),
                                    Shadow(
                                      offset: Offset(5, 5),
                                      blurRadius: 8.0,
                                      color: Color.fromARGB(125, 0, 0, 255),
                                    ),
                                  ],
                                ),
                              ),
                            ),
                          ),
                        ),
                      ),
                    ),
                  ),
                ),
              ],
            ),
            bottomNavigationBar: Container(
              height: navBarHeight,
              width: MediaQuery.of(context).size.width,
              // Option 1: Recommended
              child: RollingNavBar.iconData(
                activeBadgeColors: <Color>[
                  Colors.white,
                ],
                activeIndex: activeIndex,
                animationCurve: Curves.linear,
                animationType: AnimationType.roll,
                baseAnimationSpeed: 200,
                badges: badgeWidgets,
                iconData: iconData,
                iconColors: <Color>[Colors.grey[800]!],
                iconText: iconText,
                indicatorColors: indicatorColors,
                iconSize: 25,
                indicatorRadius: scaledHeight(context, 30),
                onAnimate: _onAnimate,
                onTap: _onTap,
              ),

              // Option 2: More complicated, but there if you need it
              // child: RollingNavBar.builder(
              //   builder: (
              //     BuildContext context,
              //     int index,
              //     AnimationInfo info,
              //     AnimationUpdate update,
              //   ) {
              //     return builderChildren[index];
              //   },
              //   badges: badgeWidgets.sublist(0, builderChildren.length),
              //   indicatorColors:
              //       indicatorColors.sublist(0, builderChildren.length),
              //   numChildren: builderChildren.length,
              //   onTap: _onTap,
              // ),
            ),
          );
        },
      ),
    );
  }
}
